Getting the authentication flow set up

jesseobrien opened the pull request:
- /login and /logout are working
- Need to validate the user credentials properly
- Need to ensure there's no security problems with the flow

jesseobrien committed:
Getting the authentication flow set up

- /login and /logout are working
- Need to validate the user credentials properly
- Need to ensure there's no security problems with the flow

jesseobrien committed:
Login/log out flow are working properly.

awly commented on internal/http/authentication.go:
```diff
@@ -0,0 +1,107 @@
+	"golang.org/x/crypto/bcrypt"
+)
+
+const AUTH_COOKIE_NAME = "user_token"
```
nit: don't use CAPITAL_CASE for global vars in Go, use regular CamelCase

awly commented on internal/http/authentication.go:
```diff
@@ -0,0 +1,107 @@
+// Store user tokens in memory.
+// Note: Nominally these would be in a storage location like redis or something
+// to be able to query for them across multiple services
+var USER_TOKENS = map[string]string{}
```
don't use global state if possible.
pass the token store around as a value

awly commented on internal/http/authentication.go:
```diff
@@ -0,0 +1,107 @@
+
+func (e ErrInvalidAuthToken) Error() string {
+	return "auth token is invalid"
+}
```
```suggestion
var ErrInvalidAuthToken = errors.New("auth token is invalid")
```

awly commented on internal/http/authentication.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+func hashUserPassword(password string) ([]byte, error) {
+	hashedpw, err := bcrypt.GenerateFromPassword([]byte(USER_PASSWORD), 10)
```
```suggestion
	hashedpw, err := bcrypt.GenerateFromPassword([]byte(password), 10)
```

awly commented on internal/http/authentication.go:
```diff
@@ -0,0 +1,107 @@
+func createTokenCookie(email string) (*http.Cookie, error) {
+
+	token := make([]byte, 26)
+	_, err := rand.Read(token)
```
```suggestion
	_, err := ioutil.ReadFull(rand.Reader, token)
```

jesseobrien replied:
I don't know that this adds anything since `rand.Read` is a wrapper around `io.ReadFull`. https://golang.org/src/crypto/rand/rand.go?s=809:847#L13

Can you elaborate on this change?

awly replied:
Ah, you're right, confused this with `rand.Reader.Read()`

awly commented on internal/http/authentication.go:
```diff
@@ -0,0 +1,107 @@
+	cookie.Expires = time.Now().Add(30 * time.Hour)
+
+	// Ensure the cookie only works over HTTPS
+	cookie.Secure = true
```
are there any other cookie properties worth setting?

jesseobrien replied:
I can set `SameSite` to `Strict` probably. I'm not sure that I want to set `HttpOnly` as then I'll need to store yet another bit in the front-end or a cookie to detect the signed in state since you can't access the cookie from `document.cookie` if `HttpOnly` is set to true.

awly commented on internal/http/http.go:
```diff
@@ -10,9 +10,18 @@ import (
+
+// As in the design doc, I'm hard coding the password and will hash it on startup. Ideally
+// the bcrypt hash would be stored in the database along with the username.
+var USER_PASSWORD = "p@ssw0rd"
```
same as above - don't use CAPITAL_CASE

awly commented on internal/http/http.go:
```diff
@@ -24,9 +33,19 @@ func Run(listenAddress, certPath, keyPath, apiToken string) error {
+	e.POST("/login", UserLoginHandler(USER_EMAIL, hashedpw))
+	e.POST("/logout", UserLogOutHandler())
+
+	e.GET("/*", echo.WrapHandler(http.StripPrefix("/", h)))
```
Check the auth cookie on these requests

russjones replied:
I think this just returns the Javascript application, so it doesn't need an auth wrapper.

awly commented on internal/http/login.go:
```diff
@@ -0,0 +1,68 @@
+		}
+
+		if err := checkEmailAndPassword(loginRequest.Email, loginRequest.Password, expectedEmail, expectedPassword); err != nil {
+			log.Printf("error user attempted to log in with invalid credentials\n")
```
```suggestion
			log.Printf("error user attempted to log in with invalid credentials: %v", err)
```

awly commented on internal/http/login.go:
```diff
@@ -0,0 +1,68 @@
+		cookie, err := createTokenCookie(loginRequest.Email)
+
+		if err != nil {
+			log.Printf("errro generating cookie for user: %v", err)
```
```suggestion
			log.Printf("error generating cookie for user: %v", err)
```

jesseobrien committed:
Update internal/http/authentication.go

Co-authored-by: Andrew Lytvynov <andrew@awly.dev>

jesseobrien committed:
Update internal/http/login.go

Co-authored-by: Andrew Lytvynov <andrew@awly.dev>

jesseobrien committed:
Update internal/http/login.go

Co-authored-by: Andrew Lytvynov <andrew@awly.dev>

jesseobrien committed:
For some reason main.go was missing...

jesseobrien committed:
Refactoring to address comments.

- Removing capitalized names
- Refactoring to pass a TokenStore around
- Adding SameSite strict mode on the cookie
- Added authentication middleware to check auth tokens on routes
- Will log user out if they try to log out and the token is already
invalid

russjones commented on internal/http/authentication.go:
```diff
@@ -0,0 +1,106 @@
+// Create a login cookie with a new token
+func createTokenCookie(tokenStore TokenStore, email string) (*http.Cookie, error) {
+
+	token := make([]byte, 26)
```
Nothing wrong with 26 bytes of random, but why specially 26? Would expect 16 or 32.

jesseobrien replied:
Hah, just popped out of my head at the time. I'll set it to 32 :+1:

russjones commented on internal/http/authentication.go:
```diff
@@ -0,0 +1,106 @@
+		return nil, err
+	}
+
+	newToken := fmt.Sprintf("%x", token)
```
```go
newToken = hex.EncodeToString(token)
```

russjones commented on internal/http/login.go:
```diff
@@ -0,0 +1,54 @@
+
+		if err := ctx.Bind(&loginRequest); err != nil {
+			// Log out the error to the logger to capture the actual error
+			log.Printf("error on user login %v\n", err)
```
Minor: `log.Printf` does not require you to add `\n`, it will add it for you if it's missing.

russjones commented on internal/http/login.go:
```diff
@@ -0,0 +1,54 @@
+		// If the user credentials are empty
+		if loginRequest == (UserLoginRequest{}) {
+			return echo.NewHTTPError(http.StatusUnauthorized, "you must provide a username and password")
+		}
```
Is this needed? The next function `checkEmailAndPassword` should fail with empty credentials anyway.

jesseobrien replied:
You're right, yeah. It's not specifically needed. I had it stubbed in before checking the email/password was there.

jesseobrien committed:
Small fixes

- Removing empty check on the credentials as there's logic that will
fail if they're empty

alex-kovoy commented on internal/http/http.go:
```diff
@@ -10,9 +10,20 @@ import (
 func Run(listenAddress, certPath, keyPath, apiToken string) error {
 	e := echo.New()
 
+	e.Use(middleware.CORS())
```
`e.Use(middleware.CORS())`

is it for development only? If so I would leave a comment.

alex-kovoy commented on web/public/app.js:
```diff
@@ -1,18 +1,60 @@
 
+axios.defaults.withCredentials = true;
+
+function getCookie(key) {
```
I do not understand the reason for giving Javascript access to cookies and a need for clearing the cookies from Javascript. Why not use httpsOnly flag and delete the cookie on the server during logout which you already do anyways?

Here is what I would do:
1) Make cookies `httpsOnly` on the server side
2) Make the LoginHandler to return session specific information (accountId, planId, etc. You can also return an user token the same that you set in the cookie and pass it with every backend call for xcsrf protection)
3) Store user session information in the `localStorage`.

Another alternative would be to remove the cookies all together and make your `LoginHandler` return an access token. The client will store this in the localStorage. This will be the same as you currently have with much less code because you do not need to deal with the cookies at all. The only difference would be that redirects will be done by the client :)

jesseobrien replied:
I agree. It's a bit sloppy to be deleting the cookies manually in the front-end. I could send back the cookie expiry on every 401 call to clear it from the server-side which would be a better flow. However, I think I'll go the last route you mentioned and use the token flow.

alex-kovoy commented on internal/http/login.go:
```diff
@@ -0,0 +1,49 @@
+			return echo.NewHTTPError(http.StatusUnauthorized, "username or password is incorrect")
+		}
+
+		cookie, err := createTokenCookie(tokenStore, loginRequest.Email)
```
(minor) `cookie` is a browser abstraction so I would name a user session something like `UserSession` or just `Session`

awly commented on cmd/fake-iot/main.go:
```diff
@@ -0,0 +1,33 @@
+	flag.StringVar(&httpListenHost, "host", "127.0.0.1", "-host=127.0.0.1")
+	flag.StringVar(&httpListenPort, "port", "8080", "-port=8080")
+	flag.StringVar(&certPath, "cert", "certs/server.crt", "-cert=path/server.crt")
+	flag.StringVar(&keyPath, "key", "certs/server.key", "-key=path/server.key")
```
nit: use actual sentences for flag descriptions (the last argument)

awly commented on internal/http/http.go:
```diff
@@ -10,9 +10,20 @@ import (
+// in the database and looked up on request
+const UserEmail = "test@example.com"
+
+type TokenStore map[string]string
```
This store needs some synchronization to avoid data races

jesseobrien committed:
Reworking auth to use token and adding CSRF middleware

- Authentication now works based on a token exchange instead of cookies.
- Token is stored in localStorage instead of cookies.
- Moved login funcs to async/await to avoid promise hell.

jesseobrien committed:
Fixing missing database

jesseobrien committed:
Removing superfluous CSRF middleware

jesseobrien committed:
Making a properly thread safe token storage implementation.

alex-kovoy reviewed (approved):
I would keep the cookie since you already had it as it adds additional depths of protection (you could do initial redirects from the server side when needed or make decisions whether to serve the static files, etc) but this solution looks good too because of simplicity :-)
